Author: Apollon Oikonomopoulos <apoikos@debian.org>
Description: Manually cleanup OpenSSL on dcrypt_deinitialize
 OpenSSL 1.1 features a cleanup function that is automatically run on shutdown
 using atexit(3). This function frees all OpenSSL-allocated resources.
 .
 In dovecot, OpenSSL is loaded indirectly using dlopen(3) against the relevant
 dovecot crypto module and is subsequently unloaded using dlclose(3). Until
 OpenSSL 1.0.1c this worked fine, however OpenSSL 1.0.1c makes sure[1] that the
 library stays loaded so that the atexit(3) handlers can run on shutdown. This,
 together with the fact that dovecot uses custom allocation functions for
 OpenSSL and has already partially freed some of OpenSSL's resources, leads to
 a segfault at process shutdown.
 .
 We fix this by explicitly calling OPENSSL_cleanup() during module unload. This
 is safe to do, as long as we will never want to subsequently re-initialize
 OpenSSL.
 .
 [1] https://github.com/openssl/openssl/commit/4af9f7fe79ff82b90c16969b7e5871435056377b
Forwarded: http://www.dovecot.org/list/dovecot/2016-November/106099.html
Last-Update: 2016-11-15
--- a/src/lib-ssl-iostream/dovecot-openssl-common.c
+++ b/src/lib-ssl-iostream/dovecot-openssl-common.c
@@ -101,6 +101,9 @@
 	ERR_remove_thread_state(NULL);
 #endif
 	ERR_free_strings();
+#if OPENSSL_VERSION_NUMBER >= 0x10100000L
+	OPENSSL_cleanup();
+#endif
 	return FALSE;
 }
 
